### PR TITLE
Introduce time-dependent inhomogeneous boundary conditions

### PR DESCRIPTION
**Pull request by mikaem**
## Summary

Inhomogeneous boundary conditions could only be set once, at construction. `DirectSumTPS.__init__` computed the lifting coefficients and froze them into `bndvals`, `inner()` collapsed the boundary block into a load vector at assembly time, and the integrators rejected time-varying boundary data outright. 

This PR makes boundary data a function of time as well as space — `u(±1,t) = g(t)` in 1D, `u(x,±1,t) = g(x,t)` in 2D — with the lifting recomputed at every Runge-Kutta stage inside `jit`.

With `u = u_h + B(t)` the weak form gains two terms:

```
M du_h/dt = A u_h + f  +  a(v, B(t))  -  <v, dB/dt>
```

`a(v, B(t))` is what `inner()` already assembled and collapsed; `<v, dB/dt>` is the term the old code was entitled to discard because `dB/dt` was always zero.

It also fixes one bug that predates the feature but only becomes reachable with it: an equation coupled to another field read that field's boundary lifting at the time the space was built, never at the stage time, so a moving wall on a *foreign* field was silently frozen.

## Types of Changes

- [x] Bug fix
- [x] New feature
- [ ] Performance improvement
- [ ] Documentation update
- [x] Refactor / cleanup
- [ ] Build / CI
- [ ] Other (describe):

## Related Issues

Fixes # \
Refs #

## Description of Changes

**The lifting becomes a plan, not a value.** `DirectSumTPS.__init__`'s four-branch construction loop became `_lifting_plans` plus a `BoundaryLifting` object: `lifting(t)` rebuilds the coefficients and `lifting.rate(t)` gives `dB/dt`. `bndvals` stays a plain dict and every existing reader is untouched; `update_bndvals(t)` refreshes it outside `jit`, while `t=`/`lifting=` on the transforms rebuild without touching the cache, which is what makes them safe under a trace.

**`dB/dt` is taken symbolically, not by AD.** The lifting is linear in the ordered boundary values, so `d/dt lift(t) = lift(dvalues/dt)(t)` exactly — the sympy expressions are differentiated and pushed through the identical plan. This avoids AD through `DiaMatrix.solve` and the FFT paths, fails loudly on what SymPy cannot differentiate where `jax.jvp` through a lambdified `Heaviside` would silently return zero, and makes the old invariant structural: a constant boundary value differentiates to exactly `0`, so the mass term vanishes symbolically rather than by assertion.

**Assembly can defer the boundary contraction.** `inner_boundary(expr)` returns a `BoundaryForcing` holding the uncontracted operators. The assembled boundary *matrices* are time-independent — the `BCGeneric` basis functions never move — so only the small coefficient array depends on `t`. A stage costs one lambdified evaluation, one forward transform per BC slot, and two matvecs. `inner`'s three contraction sites and `BoundaryForcing` now share one definition of each contraction, so agreement between the collapsed and deferred paths is structural rather than tested-into-existence.

**The integrators carry a time.** `_advance` threads a start time, `forcing_at(t)` adds `a(v,B(t)) - <v,dB/dt>`, and `IMEXRungeKutta.stage` replaces the row-sum shortcut `dt*c_i[i]*f` with `Σⱼ a_i[i][j]·f(t + c_i[j]·dt)` when the forcing moves. Steady problems keep the old branch verbatim. Constraints are algebraic, so they take no `dB/dt` term, but their lifting is still read at the stage time rather than the step time.

**Boundary data is now written in the coordinate system's own symbols.** `DirectSumTPS.__init__` used to rewrite plain sympy symbols into base scalars via `expr_psi_to_base_scalar`. That shim is gone and every example and test now builds its system first (`R(dim)`), takes `base_scalars()`/`base_time()` from it, and writes boundary values and manufactured solutions in those. See Breaking Changes.

**A coupled foreign field's lifting is deferred too.** An equation coupled to *another* equation's field gets that field's boundary lifting through `assemble_field_couplings`, not through its own trial space — and `_has_transient_boundary` only ever inspects the equation's own space. Each coupling now carries a `BoundaryForcing` evaluated at the stage time, sharing one `t_i` variable with the `resolve_constraints` call beside it so the two halves of a field cannot be read at different instants. See Screenshots / Logs.

**`forcing_at` was gated on the wrong block.** `a(v, B(t))` and `-<v, dB/dt>` come from different terms — the linear spatial one and the time derivative — and are independent, but `forcing_at` returned early whenever the *linear* block was absent. An equation with no linear spatial term therefore lost the rate term: silently, giving a 49% error, when a steady source kept `linear_forcing` non-`None`, and as a `ValueError` from `jnp.asarray(None)` when it did not. It now sums whichever blocks exist.

Two latent bugs fixed along the way. `DirectSum`'s seven transforms were `@jax.jit(static_argnums=0)` reading the lifting from *inside* the trace; since the cache keys on identity and `DirectSumTPS.__init__` reassigns `bcs` on its copies, a changed lifting was silently invisible. And `KMM2D._step_impl` did not accept the new stage time, which `ChannelFlow2D` could not have surfaced because it is excluded from the demo suite and its only runnable subclass is float64-only.

## Screenshots / Logs (if applicable)

Convergence order in `dt` with a moving wall — order, not an error threshold, because a frozen lifting still converges in space and would pass a tolerance check while ignoring time entirely:

| scheme | design order | measured |
|---|---|---|
| `ARK4_3_6L2SA` | 4 | 4.00 |
| `ARS443` | 3 | ~3 |
| `IMEX_EULER` | 1 | 1.02 |
| `BackwardEuler` | 1 | 1.00 |

Holds in 1D, in 2D with one inhomogeneous axis, and in 2D with both (the corner block). Two of the three IMEX final-combination paths are covered with a moving forcing — `full_gsa` and `implicit_only_sa`. The third is not; see Additional Notes.

```
1862 passed, 81 skipped        # uv run pytest --float64 -m "examples or not examples"
80 passed, 1 skipped           # uv run pytest -m spmd --num-devices=2   (and =4)
1189 passed, 9 skipped         # uv run pytest            (float32 default markers)
All checks passed!             # ruff check / ruff format --check / ty check
```

The coupled-field fix, on a problem where `u` is driven only through `v` and `v`
only by a wall decaying like `exp(-t)`, so `u` must decay too:

| | frozen coupling | deferred coupling |
|---|---|---|
| t=1 | 0.607 | 0.349 |
| t=5 | 0.962 | 0.032 |
| t=10 | **0.969** | **4.2e-4** |

Held at `B(0)` the coupling is a source that never switches off, and `u` climbs
to a steady state of its own instead of following the wall down.

## Breaking Changes

- [x] Yes (describe below)
- [ ] No

**1. Boundary values must use the coordinate system's symbols.** Previously a value could be written with a plain `sympy.Symbol` and `DirectSumTPS.__init__` rewrote it. Now:

```python
# before
from jaxfun.coordinates import x, y

ue = sp.exp(-(x**2 + y**2))
bcsx = {"left": {"D": ue.subs(x, 0)}, "right": {"D": ue.subs(x, 1)}}
Dx = FunctionSpace(M, space, bcs=bcsx)
...
ue = T.system.expr_psi_to_base_scalar(ue)  # converted after the fact

# after
from jaxfun.coordinates import R

C = R(2)
x, y = C.base_scalars()
ue = sp.exp(-(x**2 + y**2))
bcsx = {"left": {"D": ue.subs(x, 0)}, "right": {"D": ue.subs(x, 1)}}
Dx = FunctionSpace(M, space, bcs=bcsx, system=C)
```

`_validate_bc_symbols` enforces this at construction, comparing the symbols themselves rather than their names: a boundary value's free symbols must be among `set(system.base_scalars()) | {system.base_time()}`. Every example and test in the repo is migrated, and the suite passing is itself the check.

**2. `TimeStepper._step_impl` takes a trailing positional `t`.** Any out-of-tree subclass must add it (`t: Array | float = 0.0`, positional-only). `step`, `linear_rhs` and `total_rhs` gained `t` with defaults, so their callers are unaffected.

**3. `ETDRK4` raises `NotImplementedError` on time-dependent boundary data.** Its φ-propagators fold the forcing into the matrix exponential at setup, which is valid only for a constant forcing; carrying a moving one properly needs φ-weights against divided differences at the stage times. Reusing the constant-forcing weights would silently drop the scheme to first order, so it refuses rather than integrating a stale lifting. Use `IMEXRungeKutta` or `BackwardEuler`.

**4. Batched transforms on `DirectSumTPS` require `t` when the lifting moves.** `backward_batch`/`forward_batch`/`backward_primitive_batch` take one time per field, shape `(c.shape[0],)`. `t=None` is refused for a transient lifting rather than silently lifting every snapshot at the construction time — a batch is usually a run's snapshots, where that is quietly wrong. Steady spaces are unaffected.

**5. Pinning values are no longer coerced to Python scalars** in `Matrix`/`DiaMatrix` constraint handling, so a traced value reaches the solver instead of raising from `float()`.

**6. `examples/poisson2D_periodic_shm.py` removed**, merged into `examples/poisson2D_periodic.py`; the examples tree is now grouped by topic (`galerkin/`, `transient/`, `curvilinear/`, `navierstokes/`, `nonlinear/`, `pinns/`).

**7. `get_CoordSys` removed `cartesian_name`** as a keyword argument. Considered to be irrelevant.

## Checklist

- [x] Tests added or updated
- [x] Documentation updated (examples and docstrings; README not touched)
- [x] Added type hints
- [x] Linting & formatting pass locally (`ruff check`, `ruff format --check`, `ty check`)
- [x] All new and existing tests pass (`uv run pytest`)
- [ ] Coverage does not decrease — measured, and it drops 1 point (see below)
- [ ] Git history is clean (squash/fixup before merge)

Measured like-for-like, same command on both
(`pytest --float64 -m "examples or not examples" --cov=jaxfun`):

| module | `main` | this branch |
|---|---|---|
| `galerkin/composite.py` | 89% | 90% |
| `galerkin/inner.py` | 92% | 92% |
| `galerkin/tensorproductspace.py` | 88% | 88% |
| `integrators/base.py` | 87% | 88% |
| `integrators/constraint.py` | 86% | 86% |
| `integrators/system.py` | 95% | 95% |
| `integrators/_utils.py` | 91% | 91% |
| `integrators/imex_rk.py` | 96% | **90%** |
| **TOTAL** | **85%** | **84%** |

The box is left unticked because the total does fall. The whole of the drop is
`imex_rk.py`, and it is one identifiable thing: the *general* final-combination
path's moving-forcing branch (lines 211-213, and its system twin at 382-384)
never executes. That is new code with no test, for the reason in Additional
Notes. Every other rewritten module holds or improves. ~1700 lines of new tests
accompany ~1600 lines of new source.

New tests: `tests/galerkin/test_boundary_lifting.py` (15), `tests/galerkin/test_inner_boundary.py` (7), `tests/integrators/test_transient_bcs.py` (11), `tests/integrators/test_transient_bcs_spmd.py` (9), `tests/integrators/test_constraint_spmd.py` (4), `tests/la/test_tpmatrices_solvers_spmd.py` (5). New examples: `transient/diffusion1D_moving_wall.py`, `transient/transient2D_moving_bcs.py`.

## Additional Notes

**What guards the old behaviour.** Every new branch is gated on `is_transient` with the existing path kept verbatim beside it. `bndvals` was checked bit-for-bit against a pre-change baseline for every static fixture at each step, and `inner_boundary(expr)()` equals `inner`'s collapsed boundary vector exactly in 1D and in all four nD branches. `RayleighBenard.py`'s round-off assertion and `ChannelFlow2D` are unchanged.

**Worth a reviewer's attention.** The two-inhomogeneous-axis corner path (`projected_bcs`) is the highest-risk branch — it was untested before this PR, and it is the one whose failure mode under a stray plain symbol is silent rather than loud. `_validate_bc_symbols` now closes that.

Only boundary values a *caller* wrote are validated. `_lifting_plans` builds an inner `TensorProduct` as a projection target for the 3D `2 other + 1 bc` branch, carrying values it derived from already-validated ones and filed under a coordinate system of its own (`CartCoordSys` gives its scalars a fresh `_id`, so `R(3).x != R(2).x`). Re-validating there would compare the outer system's symbols against the inner system's and reject what the caller got right, so that construction passes `_validate_bcs=False`. The same applies to the differentiated plans behind `lifting.rate`.

**The one gap a reviewer should weigh.** `IMEXRungeKutta` has three final-combination paths. Two are covered with a moving forcing; the general one — neither globally nor implicitly stiffly accurate — is not, and its forcing branch is the only uncovered new code in this PR. The reason is that `IMEX_SSP2_222` is the sole tableau that reaches it, and it is unusable as a convergence probe here: on stiff diffusion it is stability-limited (steady walls, `|u_hat|max` at 8/32/64 steps = 6.5, 167, 0.30 against a true 0.30), and made non-stiff it converges at order **1.00** despite being a second-order scheme.

That order reduction is **not** caused by this PR. With homogeneous walls — no boundary forcing at any point — `main` and this branch produce byte-identical errors `1.79e-04, 8.92e-05, 4.46e-05, 2.23e-05`, order 1.00, while `ARS222` (`full_gsa`) and `ARK2_1_3L2SA` (`implicit_only_sa`) both give a clean 2.00 on the same problem. It is pre-existing and orthogonal, but it is why the general path has no moving-forcing test, and fixing it would let one be written.

**Deliberately out of scope.** Time-dependent *source* terms `f(x,t)*v` are still assembled once and would silently freeze. `forcing_at(t)` is the machinery that would fix them, but it changes `_prepare_inner_context`'s lambdify path. `RK4` threads `t` but has no moving-wall test.

**Where the coupling fix does *not* need a guard.** `ETDRK4` cannot silently freeze a coupling's lifting: a standalone stepper given `explicit_trials` raises `ValueError: Linear couplings to other fields need field_order` at construction, and `nonlinear_rhs` refuses couplings outright. `KMM2D` uses no coupling machinery at all. Both were checked rather than assumed, so no guard was added for either.

**Performance.** The lambdified boundary evaluation runs at trace time, not per step — `solve()` issues the same 13 `sympy.lambdify` calls whether it takes 100 steps or 400. No caching was added because none is needed on that path.

### CODE REVIEW

File: src/jaxfun/galerkin/tensorproductspace.py [hunk 1]
```diff
@@ -1459,3 +1486,306 @@ def from_orthogonal(self, c: Array) -> Array:
...
+        for val in bcs.orderedvals():
+            unknown = {str(s) for s in sp.sympify(val).free_symbols} - allowed
+            if unknown:
+                raise ValueError(
+                    f"Boundary value {val} of {space.name} uses "
+                    f"{sorted(unknown)}, which the space cannot evaluate. "
+                    f"Boundary data may depend on the coordinates "
+                    f"{sorted(allowed - {time.name})} and on time "
+                    f"({time.name}), and on nothing else."
+                )
```
[the remark this answers is not in the document]
Comment (mikaem) [reply src/jaxfun/galerkin/tensorproductspace.py:1793]: Funny the copilot should mention this, because I had already realised the problem and had it fixed, but forgot to push the commit.

File: examples/galerkin/helmholtz2D.py [hunk 2]
```diff
@@ -35,9 +33,6 @@
 v = TestFunction(T, name="v")
 u = TrialFunction(T, name="u")
 
-# Method of manufactured solution
-ue = T.system.expr_psi_to_base_scalar(ue)
```
Comment (augustfe) [inline examples/galerkin/helmholtz2D.py:39]: Nice that this is no longer necessary

File: src/jaxfun/galerkin/inner.py [hunk 3]
```diff
@@ -1029,20 +1276,46 @@ def assemble_multivar(
...
+def _forward_at(
+    V: FunctionSpaceType, uj: Array, t: float | Array | None
+) -> Array | tuple[Array, ...]: ...
+def _forward_at(
+    V: FunctionSpaceType, uj: Array, t: float | Array | None
+) -> Array | tuple[Array, ...]:
+    """Forward transform, carrying `t` only to a space that lifts boundary data."""
+    if t is None or not isinstance(V, DirectSum | DirectSumTPS):
+        return V.forward(uj)
+    return V.forward(uj, t=t)
```
Comment (augustfe) [inline src/jaxfun/galerkin/inner.py:1291]: I don't think I'm a big fan of this design. It feels more like behavior which should be guarded inside `V.forward`, but I don't immediately see a better solution than this. It's just being used within `project` and `project1D`, correct?

File: src/jaxfun/galerkin/tensorproductspace.py [hunk 4]
```diff
@@ -1119,183 +1123,53 @@ def __init__(
...
+        # is what everything downstream still reads.
+        keys = tuple(self.tpspaces)
+        time = system.base_time()
+        self.lifting = BoundaryLifting(
+            _lifting_plans(basespaces, keys, system, validate=_validate_bcs),
+            time,
+            lambda: _lifting_plans(
+                _differentiated(basespaces, time), keys, system, validate=False
+            ),
+        )
```
Comment (augustfe) [inline src/jaxfun/galerkin/tensorproductspace.py:1161]: This is a very cool design

**Review by augustfe — APPROVED** [review]
Looks very cool! So now you can implement the OpenAI Navier--Stokes solution?

File: src/jaxfun/galerkin/inner.py (hunk 3 above)
Comment (mikaem) [reply src/jaxfun/galerkin/inner.py:1291]: It's just inside `inner`. Problem is only DirectSum and DirectSumTPS have time-dependent forward(..., t=...), because of the time dependent boundary condition. Alternatively, we could add som forward(..., **kwargs) to the regular forward of all other spaces, where kwargs will soak up t=... But I don't really like that either. So since `_forward_at` is only inside `inner` I think this is a good solution.
